### PR TITLE
delta xds: fix deletion of clusters and listeners

### PR DESCRIPTION
**Pull request by howardjohn**
Due to this bug, clusters and listeners are basically never deleted.

There are two distinct bugs here:
* Our `ResourceNames` computation was wrong, which is used to determine deletes, causing misses. I also removed the lock here, since its not needed (we are operating on a clone of the WatchedResources that we write back later
* We just did not account for inbound in delta CDS at all

Fixes https://github.com/istio/istio/issues/49814

### CODE REVIEW

**Review by ramaraochavali — APPROVED** [review]

**Comment by hzxuzhonghu** [comment]
hold for a while

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 1]
```diff
@@ -123,10 +126,15 @@ func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, upd
 	for _, c := range clusters {
 		builtClusters.Insert(c.Name)
 	}
-	finalDeletedClusters := slices.FilterInPlace(deletedClusters, func(cluster string) bool {
-		return !builtClusters.Contains(cluster)
-	})
-	return clusters, finalDeletedClusters, log, true
+	// Remove anything we built from the deleted list
+	deletedClusters = deletedClusters.Difference(builtClusters)
+	// Inbound is not handled above. Remove anything that we had before and no longer have
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:131]: why inbound is not handled above, iiuc onfiggen.buildClusters generate both inbound and outbound clusters

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:131]: Inbound deletes are not handled

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 2]
```diff
@@ -113,7 +116,7 @@ func (configgen *ConfigGeneratorImpl) BuildDeltaClusters(proxy *model.Proxy, upd
 			svcs, deleted = configgen.deltaFromDestinationRules(key, proxy, subsetClusters)
 		}
 		services = append(services, svcs...)
-		deletedClusters = append(deletedClusters, deleted...)
+		deletedClusters.InsertAll(deleted...)
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:119]: I forget why we donot cal removed clusters by  watched resources names minus clusters built

File: pilot/pkg/networking/core/v1alpha3/cluster.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:131]: Can u explain why does not 	L119 deletedClusters contain inbound cluster

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:131]: Understood, it is hidden too deeper

**Review by hzxuzhonghu — APPROVED** [review]
/hold cancel

**Review by hzxuzhonghu — COMMENTED** [review]
I feel the better fix should be in deltaServices. now what is leaking is the service port is removed,. We already have proxy。ServiceTargets， so we can depend on it to cal the deleted

**Review by jacob-delgado — APPROVED** [review]

**Pull request merged**